stdlib/pqueue.ml

```diff
@@ -0,0 +1,223 @@
+      let left = left_child i in
+      if left >= len then Dynarray.set h i x (* no child, stop *) else
+      let smallest =
+        let right = right_child i in
+        if right >= len then left (* no right child *) else
+        if lt h left right then left else right
+      in
+      let y = Dynarray.get h smallest in
+      if E.compare y x < 0 then (
+        Dynarray.set h i y;
+        sift_down h ~len smallest x
+      ) else
+        Dynarray.set h i x
+
+    let pop_min h =
+      let n = Dynarray.length h in
+      if n = 0 then raise Empty;
+      let x = Dynarray.pop_last h in
+      if n = 1 then x else (
+        let r = Dynarray.get h 0 in
+        sift_down h ~len:(n - 1) 0 x;
+        r
+      )
+
+    let pop_min_opt h =
+      let n = Dynarray.length h in
+      if n = 0 then None else
+      let x = Dynarray.pop_last h in
+      if n = 1 then Some x else (
+        let r = Dynarray.get h 0 in
+        sift_down h ~len:(n - 1) 0 x;
+        Some r
+      )
+
+    let remove_min h =
+      let n = Dynarray.length h in
+      if n = 0 then raise Empty;
+      let x = Dynarray.pop_last h in
+      if n > 1 then sift_down h ~len:(n - 1) 0 x
+
+    let copy =
+      Dynarray.copy
+
+    (* array to heap in linear time (Floyd, 1964)
+
+       many elements travel a short distance, few travel longer distances
+       and we can show that it adds to O(N) *)
+    let heapify h =
+      let n = Dynarray.length h in
+      for i = n/2 - 1 downto 0 do
+        sift_down h ~len:n i (Dynarray.get h i)
+      done;
+      h
+
+    let of_array a =
+      Dynarray.of_array a |> heapify
+
+    let of_list l =
+      Dynarray.of_list l |> heapify
+
+    let of_seq s =
+      Dynarray.of_seq s |> heapify
+
+    let iter =
+      Dynarray.iter
+
+    let fold =
+      Dynarray.fold_left
+
+    let to_seq =
+      Dynarray.to_seq
+
+  end
+
+(* We provide a simpler API, where the functor only requires the priority
+   type. This is readily obtained by using the functor above on pairs
+   (priority, element). *)
+
+module type OrderedType =
+  sig
+    type t
+    val compare: t -> t -> int
+  end
+
+module type S =
+  sig
+    type prio
+    type 'a t
+    val create: unit ->'a t
+    val length: 'a t -> int
+    val is_empty: 'a t -> bool
+    val add: 'a t -> prio -> 'a -> unit
+    val add_seq: 'a t -> (prio * 'a) Seq.t -> unit
+    exception Empty
+    val min_elt: 'a t -> prio * 'a
+    val min_elt_opt: 'a t -> (prio * 'a) option
+    val pop_min: 'a t -> prio * 'a
+    val pop_min_opt: 'a t -> (prio * 'a) option
+    val remove_min: 'a t -> unit
+    val clear: 'a t -> unit
+    val copy: 'a t -> 'a t
+    val of_array: (prio * 'a) array -> 'a t
+    val of_list: (prio * 'a) list -> 'a t
+    val of_seq: (prio * 'a) Seq.t -> 'a t
+    val iter: (prio -> 'a -> unit) -> 'a t -> unit
+    val fold: ('acc -> prio -> 'a -> 'acc) -> 'acc -> 'a t -> 'acc
+    val to_seq: 'a t -> (prio * 'a) Seq.t
+  end
+
+module Make(Prio: OrderedType) =
```

Comment:
I would add a signature constraint here `: S with prio = Prio.t` to make type-checking fail earlier (than when comparing the .mli and .ml together) if S and the functor get out of sync.

Reply:
Good point. Done.